### DISCUSSION TITLE
How to properly decorate request in TypeScript

### DISCUSSION THREAD

**Discussion by awais-shafiq** [fastify/fastify#5100] [Q&A] [answered]
I'm trying this code and it fails to compile and throws an error.
```
	server.decorateRequest('user', null);

	server.addHook('preHandler', (request, _reply, done) => {
		request.user = { id: 10001 };
		done();
	});

	server.get('/', (request, reply) => {
		console.log(request.user);
		reply.code(200).send({ success: true });
	});
```

Compile Error
```
Property 'user' does not exist on type 'FastifyRequest<RouteGenericInterface, RawServerDefault, IncomingMessage, FastifySchema, FastifyTypeProviderDefault, unknown, FastifyBaseLogger, ResolveFastifyRequestType<...>>'
```

Fastify version: `4.23.2`, `4.24.0`, `4.24.2`
Node Version: `18.17.0`
Typescript version: `5.2.2`

tsconfig.json
```json
{
  "compilerOptions": {
    
    "target": "ES2017",                         
    "module": "commonjs",                       
    "rootDir": "./src",                         
    "allowJs": false,                           
    "outDir": "./dist",                         
    "noEmitOnError": true,                      
    "preserveConstEnums": true,                 
    "esModuleInterop": true,                    
    "forceConsistentCasingInFileNames": true,   
    "strict": true,                             
    "noImplicitAny": true,                      
    "strictNullChecks": true,                   
    "strictFunctionTypes": true,                
    "strictPropertyInitialization": true,       
    "useUnknownInCatchVariables": true,         
    "alwaysStrict": true,                       
    "noUnusedLocals": true,                     
    "noUnusedParameters": true,                 
    "noImplicitReturns": true,                  
    "noUncheckedIndexedAccess": true,           
    "allowUnreachableCode": true,               
    "skipLibCheck": true,                       
    "experimentalDecorators": true
  },
  "include": ["./src/**/*.ts"],
  "exclude": ["node_modules"]
}
```

**Comment by metcoder95** [comment]
Hey!

Remember to do `interface augmentation` when doing decorations to Fastify using TS, this is an example:

```ts
declare module 'fastify' {
	interface FastifyRequest {
		user: {
			id: number
		}
	}
}
```

**Comment by awais-shafiq** [comment]
In my `/src/` folder I created `custom.d.ts` file and augmented the interface as following. I'm still getting same error.

```typescript
declare module 'fastify' {
	interface FastifyRequest {
		user: {
			id: number;
		};
	}
}
```

**Reply by metcoder95** [reply, marked as answer]
You need to do the augmentation on the same place as the application code, otherwise, it rewrites the module type definition, leading to errors. A working example might be:

```ts
import { fastify } from 'fastify';

declare module 'fastify' {
  interface FastifyRequest {
    user: {
      id: number;
    };
  }
}

const server = fastify();

server.decorateRequest('user', null);

server.addHook('preHandler', (request, _reply, done) => {
  request.user = { id: 10001 };
  done();
});

server.get('/', (request, reply) => {
  console.log(request.user);
  reply.code(200).send({ success: true });
});
```

**Comment by awais-shafiq** [comment]
Attached my `tsconfig` to original question.

**Comment by awais-shafiq** [comment]
@metcoder95 Thanks!
Your solution works.

But augmenting interface in `custom.d.ts` file also worked.
I was using ts-node to run my app. It was not loading augmented interface. The solution was passing `--files` flag to ts-node command
```json
"scripts": {
    "build": "tsc ",
    "start": "node ./dist/server.js",
    "dev": "ts-node  --files ./src/server.ts "
 },
```

I was able to figure this out when I directly tried building and it did not give any error at compile time.

Again Thanks for your help.

**Comment by truongnat** [comment]
create file fastify.d.ts in src/@types
```typescript
declare module "fastify" {
  export interface FastifyInstance<
    HttpServer = http.Server,
    HttpRequest = http.IncomingMessage,
    HttpResponse = http.ServerResponse
  > {
    blipp(): void; // declare here
  }
}
```

**Comment by jpenna** [comment]
How to make it work in the instance level? The accepted solution adds the type to all requests, even those that didn't use `decorateRequest`.

For example what they ask here: https://stackoverflow.com/questions/73056086/how-to-separately-type-fastify-decorators-in-different-plugins

**Comment by jbeckton** [comment]
I was 5 minutes away from scraping my Fastify app and switching back to Hono.js. I was just trying to add some context to my request and all the official documentation did was show me how to make things worse. The one post in this thread fixed my issue. The only reason to not use Hono is that the OpenAPI integration is horrible.

**Comment by jean-michelet** [comment]
Hi :wave:

Check out the official Fastify demo, it features a fully typed application with 100% test coverage:
👉 https://github.com/fastify/demo

Also, I recently introduced a new way of working with decorators via the `getDecorator` API availabel since 5.3. This allows you to avoid module augmentation and scope your types, making typing closer to how Fastify encapsulation work. The API is available on the Fastify instance, as well as on `Request` and `Reply`. It also includes a runtime check to ensure the decorator exists, which can help catch errors early: https://fastify.dev/docs/latest/Reference/Decorators/#getdecorator-api

Here’s a working example:

```ts
interface IUser {
  name: string
}

server.register(async function (fastify) {
  const usersRepository = fastify.getDecorator<PostgreUsersRepository>(
    'usersRepository'
  )

  fastify.decorateRequest('user', null)
  fastify.addHook('onRequest', async (req, reply) => {
    req.setDecorator<IUser>('user', { name: 'Jean' })
  })

  fastify.get('/me', (request, reply) => {
    return request.getDecorator<IUser>('user')
  })
})
```

Let me know if you have any questions or feedback!

**Reply by jbeckton** [reply]
That works nicely, no additional typescript hacks to get it working.
